plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerContentProvider.java

```diff
@@ -192,38 +199,41 @@ public void handleAdd(int index, Object element) {
 				}
 				if(child instanceof IResourceUIModel && ResourceKind.ROUTE.equals(((IResourceUIModel) child).getResource().getKind())) {
 					updateChildrenFromViewer(parent);
+					elementsToRefresh.add(parent);
 				}
 			}
-			for (Object child : added) {
-				Object parent = getParent(child);
-				Trace.debug("Explorer add: parent: {0} / child: {1}", parent, child);
-				if(child instanceof Deployment) {
-					Deployment deployment = (Deployment)child;
+			for (Object addedElement : addedElements) {
+				Object parentElement = getParent(addedElement);
+				Trace.debug("Explorer add: parent: {0} / child: {1}", parentElement, addedElement);
+				if(addedElement instanceof Deployment) {
+					Deployment deployment = (Deployment)addedElement;
 					addDeploymentListeners(deployment);
+					elementsToRefresh.add(deployment.getParent());
 				}
-				if(child instanceof IResourceUIModel && ResourceKind.ROUTE.equals(((IResourceUIModel) child).getResource().getKind())) {
-					updateChildrenFromViewer(parent);
-				}
-				//HACK to fix JBIDE-21458
-				if(newList.size() > 0 && (parent instanceof IProjectAdapter || parent instanceof Deployment)) {
-					refreshViewer(parent);
-				}else {
-					addChildrenToViewer(parent, child);
-				}
+//				if(child instanceof IResourceUIModel && ResourceKind.ROUTE.equals(((IResourceUIModel) child).getResource().getKind())) {
```

Comment:
@jcantrill That's right. I reverted the comment on lines 204-206.
To verify, I applied changes from CLI using the following command : `oc delete route -l app=jee-sample` and the route disappeared from the OpenShift Explorer view. I obtained the same result when deleting the route from the 'Routes' tab in the properties view when the service was selected.

Then I used the `oc expose service jee-sample` command to expose again the service and the OpenShift Explorer view immediately displayed the route as a qualifier/decorator of the service.